### ISSUE TITLE
Node references stored in a Resource are null when playing the scene

### ISSUE THREAD

**Issue by Fabien2S** [godotengine/godot#70610]
### Godot version

v4.0.beta10.mono.official [d0398f62f]

### System information

Win11

### Issue description

When exporting a `Node` (`@export var test_node: Node`) from a `Resource` which is inside another `Node`, the value of `test_node` is always null when playing the scene. This bug also affect C#.

https://user-images.githubusercontent.com/18100602/209597467-288b97cb-9f1f-40ec-9f65-5fac868b92c6.mp4

### Steps to reproduce

1. Create a script `TestResource` extending `Resource`.
```gdscript
extends Resource
class_name TestResource

@export var test_node: Node;
```

2. Create a script extending `Node`
```gdscript
extends Node

@export var test_resource: TestResource;

func _ready():
    assert(test_resource.test_node != null)
```

4. Create a new instance of `TestResource` from the editor and set the value of `test_node`
5. Notice the "Assertion failed" error cause by the variable `test_node` being `null` when playing the scene

### Minimal reproduction project

[SampleProject.zip](https://github.com/godotengine/godot/files/10305235/SampleProject.zip)

**Comment by heppocogne** [comment]
I think this is not a bug.

test_resource.gd is initialized before test_node.gd; Marker3D does not exist when test_resource.gd is initialized, that's why `test_resource.test_node` is null.

**Comment by Fabien2S** [comment]
One thing I forgot to clarify, the bug still occurs if the value `test_node` is a parent or a sibling of the script (instead of the child as in the example)

In either case, this looks like a bug to me : if this cannot be technically possible, I shouldn't be able to select the node from the editor

**Comment by heppocogne** [comment]
Well... I'm not sure whether this is a bug or not.
How about using NodePath to select nodes from editor?

test_resource.gd
```GDScript
extends Resource
class_name TestResource

@export var test_node_path: NodePath;
```

test_node.gd
```GDScript
extends Node

@export var test_resource: TestResource;

func _ready():
	assert(get_node(test_resource.test_node_path) != null)	# this will succeed
```

**Comment by Fabien2S** [comment]
It does work when a `NodePath` is used. Its the workaround I use in my C# project but doing so I lost all the strong typing and must use `GetNode<>()` everywhere which is far from ideal.

The bug is probably tied to the resolving of the Node. I don't really know how this work under the hood but I read in some issues that a Node variable stores some kind of metadata as the path and resolve before `_ready`, but as a `Resource` does not have this function, maybe this was never implemented for a Resource-type (which is quite a weird use case to store a Node reference in a Resource anyway as it could lead to lot of "uncertainty" (What happens if the resource is saved as a file and used elsewhere?)

**Comment by stephanbogner** [comment]
I think I have the same (or a related?) issue:

**I wanted to dynamically adjust the property list if a node is selected in the inspector** (which doesn't work due to the node being null).
I also wanted to show a dropdown with the node's methods, which also doesn't work if the node itself is null.

```GDScript
@tool
extends Node2D

@export var selected_node:Node = null:
	set(value):
		selected_node = value
		notify_property_list_changed()
	
func _get_property_list():
	var properties = []
	
	print(selected_node) # Returns <null>
	
	if selected_node:
		print(selected_node.get_method_list())
		# Add dynamic property here
	return properties
```

**Comment by SaadTheGlad** [comment]
I also have come across the same issue, cannot save AnimationPlayer to a node, it spits out this bunch of errors: 
![image](https://github.com/godotengine/godot/assets/39242060/d6b89641-9a22-4001-a2e9-075d649b0aac)
![image](https://github.com/godotengine/godot/assets/39242060/482e9a8b-ca28-4273-85e4-dc999059f58d)

I guess I'll be using a NodePath along with GetNode<> but this is not ideal.

**Comment by AThousandShips** [comment]
This is not a bug, this is unsupported behavior, and fixed with an error in 4.2:
* https://github.com/godotengine/godot/pull/82843

See related reports

**Comment by AThousandShips** [comment]
We could add a warning to 4.1 to indicate that this is unsupported

**Comment by SaadTheGlad** [comment]
> We could add a warning to 4.1 to indicate that this is unsupported, but currently exporting `Node` from a resource is not supported and won't be, and likely wasn't intended to work in the first place and was an oversight

Could I ask why it won't be? And why it wasn't intended to work? Is there some deep engine limitation or?

**Comment by AThousandShips** [comment]
See the PR and linked issue :)

**Comment by SaadTheGlad** [comment]
It didn't explain it well (or I just couldn't understand it) :(

**Comment by AThousandShips** [comment]
See [here](https://github.com/godotengine/godot/issues/82809#issuecomment-1748374568) and [here](https://github.com/godotengine/godot/issues/81224)

**Comment by SaadTheGlad** [comment]
> See [here](https://github.com/godotengine/godot/issues/82809#issuecomment-1748374568) and [here](https://github.com/godotengine/godot/issues/81224)

Thank you

**Comment by akien-mga** [comment]
So this is "fixed" in 4.2 by making this an error in GDScript so it's clear that it's not supported in #82843.
